File: validation/validation_directed_graph_t.py
#  This code generates the text files used to validate the directed_graph_t
#    class of the PGL to analyze the structure of directed, unweighted and
#    simple graph.
#
#  Author:  Antoine Allard
#  WWW:     antoineallard.info
#  Date:    August 2019
#
#
#  Copyright (C) 2019 Antoine Allard
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.


import networkx as nx
import numpy as np
import sys


# Gets the name of the file containing the edgelist.
edgelist_filename = sys.argv[1]

# Loads the edgelist using networkx and makes it a simple directed graph without self-loops.
g_dir = nx.read_edgelist(edgelist_filename, create_using=nx.DiGraph())
g_dir.remove_edges_from(g_dir.selfloop_edges())
g_dir.remove_nodes_from([v for v in nx.isolates(g_dir)])

# Creates an undirected version of the graph.
g_und = g_dir.to_undirected()

# Graph properties.
with open("validation_g_prop_networkx.dat", "w") as f:
    f.write("{:>50}".format("Number of vertices:") + " ")
    f.write("{:>50}".format(g_dir.number_of_nodes()) + " \n")
    f.write("{:>50}".format("Number of edges:") + " ")
    f.write("{:>50}".format(g_dir.number_of_edges()) + " \n")
    f.write("{:>50}".format("Number of edges:") + " ")
    f.write("{:>50}".format(g_dir.number_of_edges()) + " \n")
    f.write("{:>50}".format("Density:") + " ")
    density = g_dir.number_of_edges() / (g_dir.number_of_nodes() * (g_dir.number_of_nodes() - 1))
    f.write("{:50.3f}".format(density) + " \n")
    f.write("{:>50}".format("Average in-degree:") + " ")
    f.write("{:50.3f}".format(sum([d for (n, d) in g_dir.in_degree()]) / float(g_dir.number_of_nodes())) + " \n")
    f.write("{:>50}".format("Average out-degree:") + " ")
    f.write("{:50.3f}".format(sum([d for (n, d) in g_dir.out_degree()]) / float(g_dir.number_of_nodes())) + " \n")
    f.write("{:>50}".format("Number of reciprocal edges:") + " ")
    f.write("{:>50}".format(int(nx.overall_reciprocity(g_dir)*g_dir.number_of_edges())) + " \n")
    f.write("{:>50}".format("Reciprocity (ratio definition):") + " ")
    f.write("{:50.3f}".format(nx.overall_reciprocity(g_dir)) + " \n")
    f.write("{:>50}".format("Reciprocity (statistical definition):") + " ")
    f.write("{:50.3f}".format((nx.overall_reciprocity(g_dir) - density) / (1 - density)) + " \n")
    f.write("{:>50}".format("Number of triangles:") + " ")
    f.write("{:>50}".format(int(np.sum(np.array(list(nx.algorithms.triangles(g_und).values()))) / 3)) + " \n")
    f.write("{:>50}".format("Average undirected local clustering coefficient:") + " ")
    f.write("{:50.3f}".format(nx.average_clustering(g_und)) + " \n")
    f.write("{:>50}".format("Undirected global clustering coefficient:") + " ")
    f.write("{:50.3f}".format(nx.transitivity(g_und)) + " \n")


# Vertex properties.
with open("validation_v_prop_networkx.dat", "w") as f:
    # Writes the header.
    f.write("#")
    f.write("{:>14}".format("Vertex") + " ")
    f.write("{:>15}".format("InDeg") + " ")
    f.write("{:>15}".format("OutDeg") + " ")
    f.write("{:>15}".format("DummyInt") + " ")
    f.write("{:>15}".format("DummyDbl") + " ")
    f.write("{:>15}".format("UnLocalClust") + " ")
    f.write("\n")
    # Writes the computed values.
    UnLocalClust = nx.clustering(g_und)
    for v in sorted(g_dir.nodes(), key=lambda x: (len(x), x)):
        f.write("{:>15}".format(v) + " ")
        f.write("{:>15}".format(g_dir.in_degree(v)) + " ")
        f.write("{:>15}".format(g_dir.out_degree(v)) + " ")
        f.write("{:15d}".format(np.random.randint(1000)) + " ")
        # f.write("{:>15}".format("{:15.3f}".format(100 * np.random.random()).rstrip('0').rstrip('.')) + " ")
        f.write("{:>15}".format("{:15.6e}".format(100 * np.random.random())) + " ")
        # f.write("{:>15}".format(UnLocalClust[v]) + " ")
        # f.write("{:>15}".format("{:15.9f}".format(UnLocalClust[v]).rstrip('0').rstrip('.')) + " ")
        f.write("{:>15}".format("{:15.6e}".format(UnLocalClust[v])) + " ")
        f.write("\n")


# Triangles.
triangles = sorted([sorted(x) for x in nx.enumerate_all_cliques(g_und) if len(x) == 3])
with open("validation_triangles_networkx.dat", "w") as f:
    for t in triangles:
        for v in t:
            f.write("{:>25}".format(v) + " ")
        f.write("\n")


# Triangle spectrum.
triangle_spect = {}
triangle_spect["3cycle"] = 0
triangle_spect["3nocycle"] = 0
triangle_spect["4cycle"] = 0
triangle_spect["4outward"] = 0
triangle_spect["4inward"] = 0
triangle_spect["5cycle"] = 0
triangle_spect["6cycle"] = 0
link2type = {}
link2type["111"] = "3cycle"
link2type["222"] = "3cycle"
link2type["211"] = "3nocycle"
link2type["121"] = "3nocycle"
link2type["112"] = "3nocycle"
link2type["122"] = "3nocycle"
link2type["212"] = "3nocycle"
link2type["221"] = "3nocycle"
link2type["311"] = "4cycle"
link2type["131"] = "4cycle"
link2type["113"] = "4cycle"
link2type["322"] = "4cycle"
link2type["232"] = "4cycle"
link2type["223"] = "4cycle"
link2type["123"] = "4outward"
link2type["312"] = "4outward"
link2type["231"] = "4outward"
link2type["321"] = "4inward"
link2type["132"] = "4inward"
link2type["213"] = "4inward"
link2type["133"] = "5cycle"
link2type["313"] = "5cycle"
link2type["331"] = "5cycle"
link2type["233"] = "5cycle"
link2type["323"] = "5cycle"
link2type["332"] = "5cycle"
link2type["333"] = "6cycle"
for t in triangles:
    v0 = t[0]
    v1 = t[1]
    v2 = t[2]
    edge_01 = v1 in g_dir.neighbors(v0)
    edge_10 = v0 in g_dir.neighbors(v1)
    edge_12 = v2 in g_dir.neighbors(v1)
    edge_21 = v1 in g_dir.neighbors(v2)
    edge_20 = v0 in g_dir.neighbors(v2)
    edge_02 = v2 in g_dir.neighbors(v0)
    edge_config = ""
    if edge_01 and not edge_10:
        edge_config += "1"
    if not edge_01 and edge_10:
        edge_config += "2"
    if edge_01 and edge_10:
        edge_config += "3"
    if edge_12 and not edge_21:
        edge_config += "1"
    if not edge_12 and edge_21:
        edge_config += "2"
    if edge_12 and edge_21:
        edge_config += "3"
    if edge_20 and not edge_02:
        edge_config += "1"
    if not edge_20 and edge_02:
        edge_config += "2"
    if edge_20 and edge_02:
        edge_config += "3"
    triangle_spect[link2type[edge_config]] += 1
with open("validation_triangle_spect_networkx.dat", "w") as f:
    for t in ["3cycle", "3nocycle", "4cycle", "4outward", "4inward", "5cycle", "6cycle"]:
        f.write("{:>12}".format(t) + " ")
        f.write("{:>12}".format(int(triangle_spect[t])) + " ")
        f.write("\n")

File: validation/validation_directed_graph_t.cpp
/*
 *
 *  This code generates the text files used to validate the directed_graph_t
 *    class of the PGL to analyze the structure of directed, unweighted and
 *    simple graph.
 *
 *  Compilation: g++ -O3 validation_directed_graph_t.cpp
 *
 *  Author:  Antoine Allard
 *  WWW:     antoineallard.info
 *  Date:    August 2020
 *
 *
 *  Copyright (C) 2020 Antoine Allard
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */


 // Standard template library
 #include <iomanip>
 #include <fstream>
 #include <string>
 // Portable graph library
 #include "../src/directed_graph_t.hpp"


 int main(int argc, char const *argv[])
 {
  // Name of the file containing the edgelist.
  std::string edgelist_filename = argv[1];

  // Creates the empty graph object.
  pgl::directed_graph_t g;

  // Loads the Dummy integer vertex property.
  g.load_vertices_properties("validation_v_prop_networkx.dat", 3, "DummyInt", "DummyInt", false);

  // Loads the edgelist.
  g.load_graph_from_edgelist_file(edgelist_filename);

  // Loads the Dummy double vertex property.
  g.load_vertices_properties("validation_v_prop_networkx.dat", 4, "DummyDbl");

  // Computes the degrees.
  g.compute_degrees();

  // Computes the average in-/out-degree.
  g.compute_average_vertex_prop("in-degree");
  g.compute_average_vertex_prop("out-degree");

  // Surveys the triangles and computes the undirected local clustering coefficients.
  g.survey_triangles(true, true);

  // Computes the average undirected clustering coefficient.
  g.compute_average_vertex_prop("undir_local_clust");

  // Compiles the triangle spectrum.
  g.compile_triangle_spectrum();

  // Computes the reciprocity.
  g.compute_reciprocity();


  // Writes the graph properties into a text file.
  std::ofstream g_prop_file("validation_g_prop_pgl.dat");
  std::ios_base::fmtflags g_prop_file_flags( g_prop_file.flags() );
  g_prop_file << std::setw(50) << "Number of vertices:" << " ";
  g_prop_file << std::setw(50) << g.g_prop["nb_vertices"] << " " << std::endl;
  g_prop_file << std::setw(50) << "Number of edges:" << " ";
  g_prop_file << std::setw(50) << g.g_prop["nb_edges"] << " " << std::endl;
  g_prop_file << std::setw(50) << "Number of edges:" << " ";
  g_prop_file << std::setw(50) << g.edgelist.size() << " " << std::endl;
  g_prop_file << std::setw(50) << "Density:" << " ";
  g_prop_file << std::setw(50) << std::fixed << std::setprecision(3) << g.g_prop["density"] << " " << std::endl;
  g_prop_file << std::setw(50) << "Average in-degree:" << " ";
  g_prop_file << std::setw(50) << std::fixed << std::setprecision(3) << g.g_prop["avg_in-degree"] << " " << std::endl;
  g_prop_file << std::setw(50) << "Average out-degree:" << " ";
  g_prop_file << std::setw(50) << std::fixed << std::setprecision(3) << g.g_prop["avg_out-degree"] << " " << std::endl;
  g_prop_file.flags( g_prop_file_flags );
  g_prop_file << std::setw(50) << "Number of reciprocal edges:" << " ";
  g_prop_file << std::setw(50) << g.g_prop["nb_reciprocal_edges"] << " " << std::endl;
  g_prop_file << std::setw(50) << "Reciprocity (ratio definition):" << " ";
  g_prop_file << std::setw(50) << std::fixed << std::setprecision(3) << g.g_prop["reciprocity_ratio"] << " " << std::endl;
  g_prop_file << std::setw(50) << "Reciprocity (statistical definition):" << " ";
  g_prop_file << std::setw(50) << std::fixed << std::setprecision(3) << g.g_prop["reciprocity_stat"] << " " << std::endl;
  g_prop_file.unsetf(std::ios_base::floatfield);
  g_prop_file << std::setw(50) << "Number of triangles:" << " ";
  g_prop_file << std::setw(50) << std::fixed << std::setprecision(0) << g.g_prop["nb_triangles"] << " " << std::endl;
  g_prop_file << std::setw(50) << "Average undirected local clustering coefficient:" << " ";
  g_prop_file << std::setw(50) << std::fixed << std::setprecision(3) << g.g_prop["avg_undir_local_clust"] << " " << std::endl;
  g_prop_file << std::setw(50) << "Undirected global clustering coefficient:" << " ";
  g_prop_file << std::setw(50) << std::fixed << std::setprecision(3) << g.g_prop["undir_global_clust"] << " " << std::endl;
  g_prop_file.close();


  // Writes the vertex properties into a text file.
  std::string p[] = {"in-degree", "out-degree", "DummyInt", "DummyDbl", "undir_local_clust"};
  std::vector<std::string> props(p, p+5);
  g.save_vertices_properties("validation_v_prop_pgl.dat", props);


  // Lists the triangles into a text file.
  std::set<std::string> triangles_vertices;
  std::set<std::string>::iterator triangles_vertices_it, triangles_vertices_end;
  std::set< std::set<std::string> > triangles;
  std::set< std::set<std::string> >::iterator triangles_it, triangles_end;
  for(int t(0), tt(g.g_prop["nb_triangles"]), v0, v1, v2; t<tt; ++t)
  {
    // Identifies the vertices and sorts them.
    triangles_vertices.clear();
    triangles_vertices.insert(g.ID2Name[g.triangles[t][0]]);
    triangles_vertices.insert(g.ID2Name[g.triangles[t][1]]);
    triangles_vertices.insert(g.ID2Name[g.triangles[t][2]]);
    triangles.insert(triangles_vertices);
  }
  // Writes the triangles into the file.
  std::ofstream triangles_file("validation_triangles_pgl.dat");
  triangles_it = triangles.begin();
  triangles_end = triangles.end();
  for(; triangles_it!=triangles_end; ++triangles_it)
  {
    triangles_vertices_it = triangles_it->begin();
    triangles_vertices_end = triangles_it->end();
    for(; triangles_vertices_it!=triangles_vertices_end; ++triangles_vertices_it)
    {
      triangles_file << std::setw(25) << *triangles_vertices_it << " ";
    }
    triangles_file << std::endl;
  }
  triangles_file.close();


  // Writes the triangle spectrum into a text file.
  std::ofstream triangle_spect_file("validation_triangle_spect_pgl.dat");
  triangle_spect_file << std::setw(12) << "3cycle" << " ";
  triangle_spect_file << std::setw(12) << g.triangle_spect["3cycle"] << " ";
  triangle_spect_file << std::endl;
  triangle_spect_file << std::setw(12) << "3nocycle" << " ";
  triangle_spect_file << std::setw(12) << g.triangle_spect["3nocycle"] << " ";
  triangle_spect_file << std::endl;
  triangle_spect_file << std::setw(12) << "4cycle" << " ";
  triangle_spect_file << std::setw(12) << g.triangle_spect["4cycle"] << " ";
  triangle_spect_file << std::endl;
  triangle_spect_file << std::setw(12) << "4outward" << " ";
  triangle_spect_file << std::setw(12) << g.triangle_spect["4outward"] << " ";
  triangle_spect_file << std::endl;
  triangle_spect_file << std::setw(12) << "4inward" << " ";
  triangle_spect_file << std::setw(12) << g.triangle_spect["4inward"] << " ";
  triangle_spect_file << std::endl;
  triangle_spect_file << std::setw(12) << "5cycle" << " ";
  triangle_spect_file << std::setw(12) << g.triangle_spect["5cycle"] << " ";
  triangle_spect_file << std::endl;
  triangle_spect_file << std::setw(12) << "6cycle" << " ";
  triangle_spect_file << std::setw(12) << g.triangle_spect["6cycle"] << " ";
  triangle_spect_file << std::endl;
  triangle_spect_file.close();




  return 0;
 }
